readthedocs/core/resolver.py

```diff
@@ -0,0 +1,143 @@
+"""
+Read the Docs URL Resolver.
+
+Url Types:
+
+- Subproject
+- Subdomain
+- CNAME
+- Single Version
+- Normal
+
+All possible URL's::
+
+    Subdomain or CNAME:
+
+    /<lang>/<version>/<filename> # Default
+    /<filename> # Single Version
+    /projects/<subproject_slug>/<lang>/<version>/<filename> # Subproject
+    /projects/<subproject_slug>/<filename> # Subproject Single Version
+
+    Normal Serving:
+
+    /docs/<project_slug>/<lang>/<version>/<filename> # Default
+    /docs/<project_slug>/<filename> # Single Version
+    /docs/<project_slug>/projects/<subproject_slug>/<lang>/<version>/<filename> # Subproject
+    /docs/<project_slug>/projects/<subproject_slug>/<filename> # Subproject Single Version
+"""
+
+from django.conf import settings
+import re
+
+
+def _fix_filename(project, filename):
+    """
+    Force filenames that might be HTML file paths into proper URL's
+
+    This basically means stripping / and .html endings and then re-adding them properly.
+    """
+    filename = filename.lstrip('/')
+    filename = re.sub('index.html$', '', filename)
+    filename = re.sub('index$', '', filename)
+    if filename:
+        if filename.endswith('/') or filename.endswith('.html'):
+            path = filename
+        elif project.documentation_type == "sphinx_singlehtml":
+            path = "index.html#document-" + filename
+        elif project.documentation_type in ["sphinx_htmldir", "mkdocs"]:
+            path = filename + "/"
+        else:
+            path = filename + ".html"
+    else:
+        path = ""
+    return path
```

Comment:
My initial thought here was to put this code on the specific builder backends, maybe called as a classmethod via a `project.get_backend().fix_filename()`, or something, instead of special casing one-offs here? Only a minor design decision though, really, that may not be any more manageable.